provisioner/aws-manager.js

```diff
@@ -369,84 +317,48 @@ AwsManager.prototype.listSpotRequestIds = function() {
  */
 AwsManager.prototype.capacityForType = function(workerType, states) {
   assert(workerType);
-  var that = this;
-  var wName = workerType.workerType;
-  var capacity = 0;
   if (!states) {
     states = ['running', 'pending', 'spotReq'];
   }
+  var wName = workerType.workerType;
+  var capacity = 0;
+  var instances = this.instancesOfType(workerType.workerType);
+  var requests = this.requestsOfType(workerType.workerType);
+  // TODO: Make sure internal state is tracked!
 
-  // Find instances in the retrevied state and add them to the capacity
-  // according to their declared capacity
-  this.ec2.regions.forEach(function(region) {
-    var rState = that.getApi(region);
-
-    if (!rState[wName]) {
-      return;
-    }
-
-    var wState = rState[wName];
-
-    /* When counting capacity of each type, if we have instances
-       of a type which is no longer in the workerType definition
-       we cannot know what it's capacity is.  We'll try to get its
-       real capacity, but fall back to 1 as a safe option */
-    if (states.includes('running')) {
-      wState.running.forEach(function(instance) {
-        try {
-          capacity += workerType.capacityOfType(instance.InstanceType);
-        } catch (err) {
-          capacity += 1;
-        }
-      });
-    }
 
-    if (states.includes('pending')) {
-      wState.pending.forEach(function(instance) {
-        try {
-          capacity += workerType.capacityOfType(instance.InstanceType);
-        } catch (err) {
-          capacity += 1;
-        }
-      });
+  instances.forEach(function(instance) {
+    if (states.includes(instance.State.Name)) {
+      try {
+        capacity += workerType.capacityOfType(instance.InstanceType);
+      } catch(err) {
+        capacity++;
+      }
```

Comment:
Killing them immediately doesn't feel right to me.  If we have 1000 r3.xlarge jobs which are underway but we want to switch to r3.2xlarge *only* because we're not getting xlarge ones anymore, why would we want to kill those 1000 jobs?  That puts us 1000 capacity behind, wastes the nodes we have and feels wrong.  I think that in this case, I'd rather that the worker receive the workerType updated message as we discussed and handle killing itself.

I'd like to minimize the number of moving pieces in this code for this refactor as I'd like to make sure that it gets deployed before my development head gets too far from production.  We can definately look at this later!